### ISSUE TITLE
Linear2D and Linear3D use more memory than Linear1D

### ISSUE THREAD

**Issue by Ice-cool** [hpcaitech/ColossalAI#499]
### 🐛 Describe the bug

There is a demo code to evaluate the GPU memory when use Linear1D, Linear2D and Linear3D
```
import torch

import colossalai
import colossalai.nn as col_nn
import colossalai.nn.layer.parallel_1d._utils as col_op1d
import colossalai.nn.layer.parallel_2d._operation as col_op2d
import colossalai.nn.layer.parallel_3d._operation as col_op3d

from colossalai.core import global_context as gpc
from colossalai.context import ParallelMode
from colossalai.utils import get_current_device, print_rank_0


def print_current_memory(prefix):
    max_allocated = torch.cuda.max_memory_allocated() / 1024 / 1024
    max_reserved = torch.cuda.max_memory_reserved() / 1024 / 1024
    cur_allocated = torch.cuda.memory_allocated() / 1024 / 1024
    print_rank_0(f'{prefix}: gpu(cur-used): {cur_allocated:.2f} MB, gpu(max-used): {max_allocated:.2f} MB, '
                 f'gpu(reserved): {max_reserved:.2f} MB')


def evaluate_memory(mode='1d'):
    linear = col_nn.Linear(256, 1000000, bias=False)
    for i in range(3):
        x = torch.randn(64, 256, device=get_current_device())
        x.requires_grad = True
        if mode == '1d':
            x = col_op1d.gather_forward_split_backward(x, dim=0, parallel_mode=ParallelMode.PARALLEL_1D)
            x = col_op1d.reduce_grad(x, ParallelMode.PARALLEL_1D)
        elif mode == '2d':
            x = col_op2d.all_gather_tensor_2d(x, dim=0, parallel_mode=ParallelMode.PARALLEL_2D_ROW)
            x = torch.chunk(x, 2, dim=-1)[gpc.get_local_rank(ParallelMode.PARALLEL_2D_ROW)]
        elif mode == '3d':
            x = col_op3d.all_gather_tensor_3d(x, 0, ParallelMode.PARALLEL_3D_OUTPUT)
            x = torch.chunk(x, gpc.get_world_size(ParallelMode.PARALLEL_3D_OUTPUT),
                            dim=-1)[gpc.get_local_rank(ParallelMode.PARALLEL_3D_OUTPUT)]
        else:
            raise ValueError(f'illegal mode {mode}')
        out = linear(x)
        print_current_memory('before backward')
        out.sum().backward()
        print_current_memory('after backward')


if __name__ == '__main__':
    mode = '3d'  # change to '2d', '3d'
    gpus = 8  # change to 8 for '3d'
    CONFIG = dict(parallel=dict(
        data=1,
        pipeline=1,
        tensor=dict(size=gpus, mode=mode),
    ))
    colossalai.logging.disable_existing_loggers()
    colossalai.launch_from_torch(config=CONFIG)
    evaluate_memory(mode)
```
1. Compare Linear1D and Linear2D in 4-GPUS:
```
1D-4GPUs:
before backward: gpu(cur-used): 488.59 MB, gpu(max-used): 488.59 MB, gpu(reserved): 494.00 MB
after backward: gpu(cur-used): 732.98 MB, gpu(max-used): 977.13 MB, gpu(reserved): 986.00 MB
before backward: gpu(cur-used): 732.73 MB, gpu(max-used): 977.19 MB, gpu(reserved): 986.00 MB
after backward: gpu(cur-used): 732.98 MB, gpu(max-used): 1221.27 MB, gpu(reserved): 1232.00 MB
before backward: gpu(cur-used): 732.73 MB, gpu(max-used): 1221.27 MB, gpu(reserved): 1232.00 MB
after backward: gpu(cur-used): 732.98 MB, gpu(max-used): 1221.27 MB, gpu(reserved): 1234.00 MB

2D-4GPUs:
before backward: gpu(cur-used): 488.47 MB, gpu(max-used): 976.88 MB, gpu(reserved): 986.00 MB
after backward: gpu(cur-used): 732.67 MB, gpu(max-used): 1465.22 MB, gpu(reserved): 1970.00 MB
before backward: gpu(cur-used): 732.61 MB, gpu(max-used): 1465.28 MB, gpu(reserved): 1970.00 MB
after backward: gpu(cur-used): 732.67 MB, gpu(max-used): 1709.36 MB, gpu(reserved): 2708.00 MB
before backward: gpu(cur-used): 732.61 MB, gpu(max-used): 1709.36 MB, gpu(reserved): 2954.00 MB
after backward: gpu(cur-used): 732.67 MB, gpu(max-used): 1709.36 MB, gpu(reserved): 3692.00 MB
```
2. Compare Linear1D and Linear3D in 8-GPUS
```
1D-8GPUs
before backward: gpu(cur-used): 366.77 MB, gpu(max-used): 366.77 MB, gpu(reserved): 372.00 MB
after backward: gpu(cur-used): 489.34 MB, gpu(max-used): 733.49 MB, gpu(reserved): 742.00 MB
before backward: gpu(cur-used): 488.84 MB, gpu(max-used): 733.55 MB, gpu(reserved): 744.00 MB
after backward: gpu(cur-used): 489.34 MB, gpu(max-used): 855.56 MB, gpu(reserved): 990.00 MB
before backward: gpu(cur-used): 488.84 MB, gpu(max-used): 855.56 MB, gpu(reserved): 990.00 MB
after backward: gpu(cur-used): 489.34 MB, gpu(max-used): 855.56 MB, gpu(reserved): 992.00 MB

3D-8GPUs
before backward: gpu(cur-used): 610.66 MB, gpu(max-used): 1587.23 MB, gpu(reserved): 1722.00 MB
after backward: gpu(cur-used): 488.53 MB, gpu(max-used): 1953.50 MB, gpu(reserved): 3688.00 MB
before backward: gpu(cur-used): 732.73 MB, gpu(max-used): 1953.56 MB, gpu(reserved): 5284.00 MB
after backward: gpu(cur-used): 488.53 MB, gpu(max-used): 2075.57 MB, gpu(reserved): 7376.00 MB
before backward: gpu(cur-used): 732.73 MB, gpu(max-used): 2075.57 MB, gpu(reserved): 8972.00 MB
after backward: gpu(cur-used): 488.53 MB, gpu(max-used): 2075.57 MB, gpu(reserved): 11062.00 MB
```
### Environment

PyTorch version：1.8.1
Colossal version：0.1.0

**Comment by FrankLeeeee** [comment]
Hi, can you let know how many GPUs are you using?

**Comment by Ice-cool** [comment]
Linear1D vs Linear2D：use 4GPUs
Linear1D vs Linear3D：use 8GPUs

**Comment by kurisusnowdeng** [comment]
Could you please also print the shapes of `x` between preprocessing (those gatherings and chunkings) and computing `linear_1`? This may help us know exactly how much data is processed by each worker.

**Comment by Ice-cool** [comment]
我暂时采用中文进行回复了（沟通后后续可以改成英文）
说明：官方例子是认为每张卡上都拥有全部的数据，此处是预先假设数据Data Parallel在不同卡上
设定：每块GPU上面都有一个`64x256 (batch-size x dim)` 的输入，针对3种不同Linear情况各自进行下述操作，以保证最后的输出是等价的（上述的gather和chunk操作都是为了保证最后输出是一致的）
4GPUs情况下Linear1D vs Linear2D：此时总的数据大小为（`(64*4)x256`）
           - Linear1D对weight进行拆分，并未对input进行拆分，因此需要先gather所有数据，每块卡上的输入变为 `(64*4)x256`
           - Linear2D对weight和input进行了四宫格拆分，因此需要只需将两张卡上数据收集，然后按照dim进行砍半，此时每张卡上的理论输入是 `(64*2)x128` 

8GPUs的情况类似~

**Comment by kurisusnowdeng** [comment]
Suppose the complete input shape is [64, 256], please try the following ways to get input tensor for each worker
1d:
```python
    x = torch.randn(64, 256, device=get_current_device())
    src = gpc.get_ranks_in_group(ParallelMode.TENSOR)[0]
    torch.distributed.broadcast(x, src, group=gpc.get_group(ParallelMode.TENSOR))
```

2d:
```python
    x = torch.randn(64, 256, device=get_current_device())
    src = gpc.get_ranks_in_group(ParallelMode.TENSOR)[0]
    torch.distributed.broadcast(x, src, group=gpc.get_group(ParallelMode.TENSOR))
    x = col_op2d.split_tensor_2d(x)
    x = torch.chunk(x, gpc.get_world_size(ParallelMode.PARALLEL_2D_ROW), dim=-1)
    x = x[gpc.get_local_rank(ParallelMode.PARALLEL_2D_ROW)].contiguous()
```

3d:
```python
    x = torch.randn(64, 256, device=get_current_device())
    src = gpc.get_ranks_in_group(ParallelMode.TENSOR)[0]
    torch.distributed.broadcast(x, src, group=gpc.get_group(ParallelMode.TENSOR))
    x = col_op3d.split_batch_3d(x)
    x = col_op3d.split_tensor_3d(x, -1, ParallelMode.PARALLEL_3D_OUTPUT)
```

BTW, please reset the memory statistics before computing like
```python
    x = ...
    torch.cuda.reset_peak_memory_stats(gpc.get_global_rank())
    out = linear(x)
```
Also, please share the results with us after applying the changes above. Thank you!

**Comment by Ice-cool** [comment]
after applying the changes above, the results as follows (add `torch.cuda.reset_peak_memory_stats(gpc.get_global_rank())` before print memory information):  
4GPUs:  Linear1D vs Linear2D
```
1D-4GPUs:
before backward: gpu(cur-used): 306.20 MB, gpu(max-used): 306.20 MB, gpu(reserved): 310.00 MB
after backward: gpu(cur-used): 550.34 MB, gpu(max-used): 550.34 MB, gpu(reserved): 618.00 MB
before backward: gpu(cur-used): 550.34 MB, gpu(max-used): 550.34 MB, gpu(reserved): 618.00 MB
after backward: gpu(cur-used): 550.34 MB, gpu(max-used): 550.34 MB, gpu(reserved): 864.00 MB
before backward: gpu(cur-used): 550.34 MB, gpu(max-used): 550.34 MB, gpu(reserved): 864.00 MB
after backward: gpu(cur-used): 550.34 MB, gpu(max-used): 550.34 MB, gpu(reserved): 864.00 MB

2D-4GPUs:
before backward: gpu(cur-used): 306.16 MB, gpu(max-used): 306.16 MB, gpu(reserved): 802.00 MB
after backward: gpu(cur-used): 550.30 MB, gpu(max-used): 550.30 MB, gpu(reserved): 1602.00 MB
before backward: gpu(cur-used): 550.30 MB, gpu(max-used): 550.30 MB, gpu(reserved): 2094.00 MB
after backward: gpu(cur-used): 550.30 MB, gpu(max-used): 550.30 MB, gpu(reserved): 3078.00 MB
before backward: gpu(cur-used): 550.30 MB, gpu(max-used): 550.30 MB, gpu(reserved): 3078.00 MB
after backward: gpu(cur-used): 550.30 MB, gpu(max-used): 550.30 MB, gpu(reserved): 3816.00 MB
```
8GPUs: linear1D vs Linear3D
```
1D-8GPUs
before backward: gpu(cur-used): 152.65 MB, gpu(max-used): 152.65 MB, gpu(reserved): 158.00 MB
after backward: gpu(cur-used): 274.72 MB, gpu(max-used): 274.72 MB, gpu(reserved): 314.00 MB
before backward: gpu(cur-used): 274.72 MB, gpu(max-used): 274.72 MB, gpu(reserved): 314.00 MB
after backward: gpu(cur-used): 274.72 MB, gpu(max-used): 274.72 MB, gpu(reserved): 438.00 MB
before backward: gpu(cur-used): 274.72 MB, gpu(max-used): 274.72 MB, gpu(reserved): 438.00 MB
after backward: gpu(cur-used): 274.72 MB, gpu(max-used): 274.72 MB, gpu(reserved): 438.00 MB

3D-8GPUs
before backward: gpu(cur-used): 396.75 MB, gpu(max-used): 396.75 MB, gpu(reserved): 804.00 MB
after backward: gpu(cur-used): 274.67 MB, gpu(max-used): 274.67 MB, gpu(reserved): 1544.00 MB
before backward: gpu(cur-used): 518.82 MB, gpu(max-used): 518.82 MB, gpu(reserved): 1914.00 MB
after backward: gpu(cur-used): 274.67 MB, gpu(max-used): 274.67 MB, gpu(reserved): 2160.00 MB
before backward: gpu(cur-used): 518.82 MB, gpu(max-used): 518.82 MB, gpu(reserved): 2222.00 MB
after backward: gpu(cur-used): 274.67 MB, gpu(max-used): 274.67 MB, gpu(reserved): 3086.00 MB
```

**Comment by kurisusnowdeng** [comment]
Hi, @Ice-cool . I think the new results basically make sense. Since the weight is much larger than input, we can simply consider the memory usage of the weight tensor.

On 4 GPUs, 2D and 1D both consume 1/4 of the complete weight tensor, so the results are very close.

On 8 GPUs, 3D and 1D both store 1/8 partitions of weight (cur_used). The peak memory usage of 3D is 1/4 of weight, while 1D is 1/8 (max-used).

**Comment by Ice-cool** [comment]
thank you for your reply, i am still curious about why the max_memory_reserved in Linear2D and Linear3D is much larger than Linear1D.

**Comment by FrankLeeeee** [comment]
Hi, in fact, I found this issue as well. Usually, a large gap between allocated memory and cached memory suggests the presence of memory fragmentation. I am investigating deeper into this issue. I will update you if I reach any conclusion :)

**Comment by Ice-cool** [comment]
I agree with you. However, here is a special case in face recognition, there is a large linear layer (dim x num_classes, the num_classes is very large), the largest batch-size in Linear2D is smaller than Linear1D in same setting.
